datafusion/sql/src/planner.rs

```diff
@@ -5509,6 +5599,99 @@ mod tests {
+        let sql = "SELECT id, order_id \
+            FROM person \
+            INNER JOIN orders \
+            ON person.id + 10 = orders.customer_id * 2";
+
+        let expected = "Projection: person.id, orders.order_id\
+        \n  Inner Join: person.id + Int64(10) = orders.customer_id * Int64(2)\
+        \n    Projection: person.id, person.first_name, person.last_name, person.age, person.state, person.salary, person.birth_date, person.😀, person.id + Int64(10)\
+        \n      TableScan: person\
+        \n    Projection: orders.order_id, orders.customer_id, orders.o_item_id, orders.qty, orders.price, orders.delivered, orders.customer_id * Int64(2)\
+        \n      TableScan: orders";
+        quick_test(sql, expected);
+    }
+
+    #[test]
+    fn test_multiple_column_expr_eq_join() {
+        let sql = "SELECT id, order_id \
+            FROM person \
+            INNER JOIN orders \
+            ON person.id + person.age + 10 = orders.customer_id * 2 - orders.price";
+
+        let expected = "Projection: person.id, orders.order_id\
+        \n  Inner Join: person.id + person.age + Int64(10) = orders.customer_id * Int64(2) - orders.price\
+        \n    Projection: person.id, person.first_name, person.last_name, person.age, person.state, person.salary, person.birth_date, person.😀, person.id + person.age + Int64(10)\
+        \n      TableScan: person\
+        \n    Projection: orders.order_id, orders.customer_id, orders.o_item_id, orders.qty, orders.price, orders.delivered, orders.customer_id * Int64(2) - orders.price\
+        \n      TableScan: orders";
+        quick_test(sql, expected);
+    }
+
+    #[test]
+    fn test_left_projection_expr_eq_join() {
+        let sql = "SELECT id, order_id \
+            FROM person \
+            INNER JOIN orders \
+            ON person.id + person.age + 10 = orders.customer_id";
+
+        let expected = "Projection: person.id, orders.order_id\
+        \n  Inner Join: person.id + person.age + Int64(10) = orders.customer_id\
+        \n    Projection: person.id, person.first_name, person.last_name, person.age, person.state, person.salary, person.birth_date, person.😀, person.id + person.age + Int64(10)\
+        \n      TableScan: person\
+        \n    TableScan: orders";
+        quick_test(sql, expected);
+    }
+
+    #[test]
+    fn test_right_projection_expr_eq_join() {
+        let sql = "SELECT id, order_id \
+            FROM person \
+            INNER JOIN orders \
+            ON person.id = orders.customer_id * 2 - orders.price";
+
+        let expected = "Projection: person.id, orders.order_id\
+        \n  Inner Join: person.id = orders.customer_id * Int64(2) - orders.price\
+        \n    TableScan: person\
+        \n    Projection: orders.order_id, orders.customer_id, orders.o_item_id, orders.qty, orders.price, orders.delivered, orders.customer_id * Int64(2) - orders.price\
+        \n      TableScan: orders";
+        quick_test(sql, expected);
+    }
+
```

Comment:
Very nice test !

As I recall this type of transformation can not always be done for all types of joins (though perhaps it is ok for equijoins)

For example, I wonder if it is correct to always pull expressions below the join for FULL OUTER JOINs 🤔 

Can you add a test for something like

```
 let sql = "SELECT id, order_id \
            FROM person \
            FULL OUTER JOIN orders \
            ON person.id = 10";
```

if it doesn't already exist?

In that case the predicates need to be evaluated within the join (if it is done as a filter afterwards it may filter rows that should not be filtered)

Reply:
>  let sql = "SELECT id, order_id \
            FROM person \
            FULL OUTER JOIN orders \
            ON person.id = 10";

currently I transfer `this sql` to `cross join` as master does, but as you explained it may filter rows that should not be filtered, we should do same for this sql.

Unfortunately, one test case will fail caused by different data type of join keys, after I add the change.
https://github.com/apache/arrow-datafusion/blob/406c1087bc16f8d2a49e5a9b05d2a0e1b67f7aa5/datafusion/core/tests/sql/joins.rs#L369-L385

I think we can fix it after we add `type_coercion` for join in optimizer rule.
* https://github.com/apache/arrow-datafusion/issues/2877

Any way, I add the full join test case.

Reply:
> For example, I wonder if it is correct to always pull expressions below the join for FULL OUTER JOINs 🤔
> 
> Can you add a test for something like
> 
> ```
>  let sql = "SELECT id, order_id \
>             FROM person \
>             FULL OUTER JOIN orders \
>             ON person.id = 10";
> ```
> 
> if it doesn't already exist?
> 
> In that case the predicates need to be evaluated within the join (if it is done as a filter afterwards it may filter rows that should not be filtered)

I think it is OK to projection such kind expressions, the type of the new projected column type is bool, the bool value is 
evaluated again during the join.


One thing that I am not clear now and whether we should cover in this PR and can do in the following PR is
Should we allow the following cases ?

// Join conditions include suspicious Exprs
SELECT id, order_id 
               FROM person 
               FULL OUTER JOIN orders 
               ON person.id = person.name = order.id

 
// Join conditions include non-deterministic Exprs
SELECT id, order_id 
               FROM person 
               FULL OUTER JOIN orders 
               ON person.id = person.name || random(xxx)

Reply:
> I think it is OK to projection such kind expressions, the type of the new projected column type is bool, the bool value is evaluated again during the join.

It seems your idea make things simple.

Currently for join condition like `person.id = 10`, I treat `person.id` as left key, and `10` as `right key`, this need do type coercion.

But as you suggestion, we can treat `person.id = 10` as left key, and right key as `true`, 
then exist `type coercion` will handle this condition.

I rewrite the expression to `(person.id = 10) = true`, then the test passes now. 
The full join test case will like:
```rust
#[test]
    fn test_one_side_constant_full_join() {
        let sql = "SELECT id, order_id \
            FROM person \
            FULL OUTER JOIN orders \
            ON person.id = 10";

        let expected = "Projection: person.id, orders.order_id\
        \n  Filter: person.id = Int64(10) = Boolean(true)\
        \n    CrossJoin:\
        \n      TableScan: person\
        \n      TableScan: orders";
        quick_test(sql, expected);
    }
```